Is game in check

saakuffo opened the pull request:
**Added logic to determine if the game is in check**
- The logic is pretty straight forward. The method will be passed the opponent's color (that is, whoever turn it is NOT)
- It will use that color to query the database to get all the opponent's pieces that are NOT captured
- it will query the King of the current player (the one who's turn it is)
- It will setup a Boolean variable for in_check
- in the case of a Pawn it will call a new method to check if the pawn can capture **Note this will probably necessitate a new card to create a method to allow pawns to capture. I will do this as a separate pull request.**
- if the king is in a valid_pawn_capture location the game is in check
- For all other opponents pieces, it will check if the the piece has a valid move on the King's X and Y co-ordinates
- If there is a valid move, the game is in check
- If the king's x and y location is not the location of a valid_move? for all non-pawn pieces and not a valid_pawn_capture? location the game is **NOT in check**

**Added the general valid_pawn_capture logic**
- This was bundled with this PR because it was required for a fully functioning method. In the future this would be a separate PR
- The method defaulty falls through to returning false
-  pawns can only capture diagonally. In the future, we can substitute this check for the is_obstructed method diagonally, but this will be a temporary solution
- If a piece is one spot to the left or one spot to the right (x == current_piece.x -1 or x ==current_piece.x  +1) check for color of the piece
- If the color of the piece is black, check to see if the y location is below the current piece (if y == current_piece.y + 1)
- else check to see if the y location is above the current piece (if y == current_piece.y - 1)
- if either of those are true, the piece is capturable by the pawn.
- else, as noted above, return false

**Factories and tests updates**
- wrote tests for the above 2 in the context of is game in check (can a rook capture a king and can a pawn capture a king
- wrote skeleton factories for all pieces. This is mean to be easy to use factories to replace some of the manual piece logic we are writing in tests right now. **Dave, please let me know your thoughts on this because it seemed a tad superfluous**

saakuffo commented:
Looks like the wrong branch pushed. Fixing.

lasley commented on app/models/game.rb:
```diff
@@ -12,43 +12,63 @@ def initialize_board
-        color: 'white'
+        y_position: 6,
+        color: 'white',
+        is_captured: false
```
Superfluous is_captured - Booleans default to false if not explicitly set. Apply to all

lasley commented on app/models/pawn.rb:
```diff
@@ -6,6 +6,17 @@ def valid_move?(x, y)
 
+  def valid_pawn_capture?(x, y)
+    if x == x_position + 1 || x == x_position - 1
+      if color == 'black'
```
```
modifier = (color == 'black') ? 1 : -1
return y == (y_position + modifier)
```

saakuffo replied:
Dave, not sure I follow the logic. 
Pawns don't capture up and down. They capture diagonally. 

Essentially I have to check if there is a piece diagonal to the pawn, in the direction it would be going and capture it.

This can probably be simplified with the obstructed method but still needs to worry about horizontal planes.

lasley replied:
The logic I put is a simplified form of your L11-14 from what I can see. I would love to see a [truth table](https://en.wikipedia.org/wiki/Truth_table#Binary_operations) or implementation proving me wrong though

lasley commented on app/models/pawn.rb:
```diff
@@ -6,6 +6,17 @@ def valid_move?(x, y)
   # method to define if the pawn movement is of advance
 
+  def valid_pawn_capture?(x, y)
+    if x == x_position + 1 || x == x_position - 1
```
Now that I'm looking harder, this can be simplified too:

```
if x in [x_position + 1, x_position - 1]
```

lasley replied:
Note that tradeoff is efficiency, for something this small anyways. Technically you're:
* Creating the array (too bad Ruby doesn't have tuples or other immutable iterator - this is more efficient than in better langs like Python, Go, C, Java...)
* Iterating array, if statement on each value

So you gain the overhead of creating the data structure basically. You also perform any method calls, which could be bad and is worth paying attention to. You gain scalability, maintainability, and DRY though.

saakuffo commented:
@lasley changed made. 
- Removed is_captured: false assignments removed
- simplified is pawn captured logic as suggested. Note this method is on the block for refactor so we can possibly use the obstructed method in some way.
